로그인 페이지 권한 추가, 공지사항 댓글 읽어오기 기능 추가

kingle1024 opened the pull request:
[로그인 페이지 및 권한 추가(Member 관련 파일들)]
- Security 처리 진행(AuthenticationSuccessHandler.java, SecurityConfiguration.java)
- Member Repository 추가 및 로그인 성공 시 회원 정보 가져오기 로직 추가(없으면 에러)

[공지사항 삭제 기능 Controller 추가(ApiNoticeController)]
- controller를 실수로 커밋하지 않았어서 추가 커밋 진행
- service는 이전 풀리퀘에 존재 https://github.com/kingle1024/LabManage/pull/10/commits/444aa786a1dd9b41f29ab6313979268d520f0254

[공지사항 댓글 읽어오기 기능 추가]
- 해당 boardNo에 맞는 Comment들만 가져와서 목록으로 뿌려줌 
- 테스트 코드 작성 완료 

------------------------------------------------------------------------------

반성 : 풀리퀘를 좀 더 잘라서 하겠습니다. 
push한거 모두 회수하고 다시 해볼까 했는데 꼬일까 싶어서 일단 올리고 다음거 부터 조심하겠습니다!

kingle1024 committed:
Security 로그인 페이지 및 권한 추가

kingle1024 committed:
Data가 필요한 이유 : view의 반환을 요청하는 값과 controller에서 반환해주는 타입을 일치 시키기 위해 @Data를 통해 Getter를 생성해줌

kingle1024 committed:
feat : 로그인 및 페이지 구축

kingle1024 committed:
삭제 버튼 관리자만 보이도록 처리

kingle1024 committed:
로그인, 로그아웃, 회원 관리 링크 권한에 따라 보여지도록 처리

kingle1024 committed:
feat : 공지사항 상세에 있는 댓글 읽어오기 기능 추가 Closed #11

kingle1024 committed:
refactor : JPA에서 findBy 시 에러가 나서 변수명에 언더바(_) 있는 것들 제거

kingle1024 committed:
refactor : JPA에서 findBy 시 에러가 나서 변수명에 언더바(_) 있는 것들 제거

kingle1024 committed:
Merge remote-tracking branch 'origin/dev' into dev

KimHyoJin commented on src/main/java/com/gworld/manage/comment/repository/CommentRepository.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.Optional;
+
+public interface CommentRepository extends JpaRepository<Comment, Long> {
+    Optional<List<Comment>> findAllByBoardNoAndTypeCode(long boardNo, String typeCode);
```
List는 Optional일 필요가 없을 것 같습니다-! Optional은 NPE를 막기위해 많이 사용하는데요, List의 경우 emptyList가 존재하므로 Optional로 한번 더 감싸지 않아도 될것 같아요.

kingle1024 replied:
- 컬렉션을 반환하는 Spring Data JPA Repository 메서드는 null을 반환하지 않고 비어있는 컬렉션을 반환해주므로 Optional로 감싸서 반환할 필요가 없음!
6e5b236614e261eedb004225f2ce5dbb00b5cc98

KimHyoJin commented on src/main/java/com/gworld/manage/comment/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,25 @@
+    private final CommentRepository commentRepository;
+
+    @Override
+    public List<Comment> list(Long boardNo, String typeCode) {
```
comment의 경우 굉장히 많이 달릴수가 있으므로 pagination처리를 해주는 것이 좋을 것 같습니다-!

kingle1024 replied:
넵넵, pagination은 추가 개발 후 다시 풀리퀘 하겠습니다~

KimHyoJin commented on src/main/java/com/gworld/manage/member/entity/Member.java:
```diff
@@ -0,0 +1,26 @@
+    private String profileImagePath;
+    private boolean enabled;
+    private LocalDateTime registerDate;
+    private boolean adminYn;
```
admin 여부를 나타내는 필드면 isAdmin 정도가 좀더 깔끔하지 않을까 싶어요!

kingle1024 replied:
오호 제가 봐도 isAdmin이 더 좋아보이네요. 모두 수정하겠습니다~

kingle1024 replied:
https://github.com/kingle1024/LabManage/pull/12/commits/b8561f2195231c4f634904702d5b3e4609fc1c2d

KimHyoJin commented on src/main/java/com/gworld/manage/comment/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,25 @@
+    private final CommentRepository commentRepository;
+
+    @Override
+    public List<Comment> list(Long boardNo, String typeCode) {
```
Comment가 entity로 보이는데 entity를 controller단까지 반환하는건 비추천하는 패턴입니다. layer별로 사용하는 클래스를 분리해서 사용하는 것이 좋을 것 같아요!

kingle1024 replied:
@KimHyoJin 
강의 들으면서 계속해서 클래스를 나누는 이유가 궁금했었는데, 리뷰 덕분에 나누는 이유에 대해서 찾아보고 많이 공부할 수 있었습니다!
아래와 같이 layer별로 Param, Dto로 나누어주었는데, 의도하신 말씀이 맞을까요?
CommentDto 하나로 View <--> Controller <--> Service Layer를 이동하는 것은 권장하지 않는 것 맞나요?

View <--> Controller : CommentParam
Controller <--> Service : CommentDto 

리뷰로 작성한 코드 링크
775a05dcae50d0c8097713ea9b6c921c91201e63

KimHyoJin commented on src/main/java/com/gworld/manage/configuration/AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,22 @@
+
+@RequiredArgsConstructor
+@Configuration
+public class AuthenticationSuccessHandler extends SavedRequestAwareAuthenticationSuccessHandler {
```
SecurityConfiguration 에서 받아오려면 bean으로 따로 선언해야할것 같아요!

kingle1024 replied:
@KimHyoJin 
SecurityConfiguration Bean으로 찾아보았는데, 제가 이해를 잘 못한 것 같습니다ㅜㅜ 
@Configuration에 @Component가 있고, Bean 처리를 해준다는 것까지는 찾았는데,
아래와 같이 bean 파일을 새로 만들어서 명시해 주어야 한다는 뜻일까요?
구글링 검색어나 조금만 힌트를 더 주시면 더 공부해보겠습니다 (__)


```
<beans:beans
  <http>
   ....
    <!-- 성공 핸들러 등록 -->
    <beans:bean id="AuthenticationSuccessHandler" class="com.gworld.manage.configuration.AuthenticationSuccessHandler">
        <beans:property name="id" value="id" /> 
        <beans:property name="password" value="1234" /> 
    </beans:bean> 
</beans:beans>

```

KimHyoJin replied:
앗! 제가 잘못보았던것 같습니다. 우선 말씀하신 내용이 맞아요. 
다만 위 클래스는 클래스 내부에서 @bean을 따로 선언하지 않기때문에 configuration보단 component로 선언하시면 좀더 좋을 것 같습니다. configurtion 어노테이션은 보통 다음과 같은 형태로 사용합니다 
~~~
@Configuration
   public class AppConfig {
  
       @Bean
       public MyBean myBean() {
           // instantiate, configure and return bean ...
       }
   }
 ~~~

KimHyoJin commented on src/main/java/com/gworld/manage/configuration/SecurityConfiguration.java:
```diff
@@ -1,16 +1,30 @@
 
+    private final MemberService memberService;
+    @Autowired
+    public AuthenticationSuccessHandler authenticationSuccessHandler;
```
@RequiredArgsConstructor 가 있으므로 private final로 선언해도 될 것 같습니다.

kingle1024 replied:
맞네요! 감사합니다!
dbeecd55d9ae934a81fedfe4e87f192d48d07f0a

KimHyoJin commented on src/main/java/com/gworld/manage/controller/IndexController.java:
```diff
@@ -0,0 +1,10 @@
+
+import org.springframework.web.bind.annotation.RequestMapping;
+
+public class IndexController {
```
@Controller annotation이 없어보여요..! 없어도 괜찮나요?

kingle1024 replied:
엇.. 감사합니다! 없어도 작동하긴 했었다보니 발견을 못했었네요

KimHyoJin reviewed (approved):
우와-!! 수고 많으셨습니다-!  PR크기는 제가 보기엔 적당해보입니다ㅎㅎ 커밋별로 잘 나눠져 있어서 좋았어요. 
몇가지 코멘트 드렸는데 한번 확인부탁드립니다! 수고 많으셨어요~~!!!

kingle1024 committed:
refactor : Controller Anotation 추가

kingle1024 committed:
refactor : adminYn -> isAdmin 으로 변경

kingle1024 committed:
refactor : @RequiredArgsConstructor가 있으므로 생성자를 생성해주기 때문에 @Autowired를 제거하고 private final로 선언 가능!

kingle1024 committed:
refactor : 컬렉션을 반환하는 Spring Data JPA Repository 메서드는 null을 반환하지 않고 비어있는 컬렉션을 반환해주므로 Optional로 감싸서 반환할 필요가 없음!

kingle1024 committed:
refactor : layer별로 사용하는 class 분리 처리
- View <--> Controller 통신은 CommentParam, BoardParam
- Controller <--> Service 통신은 CommentDto, BoardDto

kingle1024 committed:
feature : 에러 코드 및 핸들러 추가

kingle1024 committed:
feature : 공지사항 에러 Exception 추가

kingle1024 committed:
refactor : WebSecurityConfigurerAdapter is deprecated에 따른 구조 변환 작업 진행

KimHyoJin commented:
ㅠㅠ 제가 리뷰를 달아두었었는데 pending상태로 두었었네요! 추가했습니다. 그리고 SonarCloud까지 붙이시다니!!! 👍🏻

kingle1024 commented:
> ㅠㅠ 제가 리뷰를 달아두었었는데 pending상태로 두었었네요! 추가했습니다. 그리고 SonarCloud까지 붙이시다니!!! 👍🏻

멘토님 덕분에 코드 리뷰를 경험하다보니, 좋은 코드 작성하는 방법에 대해서 고민하다가 적용해보게 되었습니다 :)

kingle1024 committed:
all data upload
